core/lsp/helpers.cc

```diff
@@ -123,6 +126,7 @@ string prettyDefForMethod(const core::GlobalState &gs, core::MethodRef method) {
         methodNamePrefix = "self.";
     }
     vector<string> prettyArgs;
+    auto defaultArgumentPlaceholder = concretizeIfAbstract ? "T.let(T.unsafe(nil), T.untyped)" : "…";
```

Comment:
It looks like `concretizeIfAbstract` is doing double-duty here.

You're using it both to decide to convert `abstract` to `override`, but also to decide to not generate the syntactically-invalid `…` character as a default.

We should probably have a separate variable for this, and IMO, this variable already exists: the `ShowOptions` struct defined in `core/ShowOptions.h`.

As you noted, the code you're writing appears very similar to the `package/rbi_gen.cc` code. That code was written to generate RBI files from types, and when it was written, we invented this `ShowOptions` struct so we would never again have to thread through a ton of special printing args through all of our printing methods (#5102).

IMO, we should add `concretizeIfAbstract` as a boolean to `ShowOptions`, and then take `ShowOptions` in this method. That's important, because separately, you have a bug in your code for `prettySigForMethod` due to the fact that it *doesn't* take `ShowOptions`:

https://sorbet.run/#%23%20typed%3A%20true%0A%0Aclass%20A%0A%20%20extend%20T%3A%3ASig%0A%20%20sig%20%7Breturns%28T.attached_class%29%7D%0A%20%20def%20self.example%0A%20%20%20%20%23%20%20%20%20%20%20%5E%20hover%20here%0A%20%20%20%20new%0A%20%20end%0Aend

In this playground, if you hover over `example`, you'll see that Sorbet reveals

```ruby
sig {returns(T.attached_class (of A))}
```

for the sig. That is not valid syntax. The `... (of A)` suffix can be turned off by setting `showForRBI` to `true`, so long as `prettySigForMethod` takes `ShowOptions`, and passes it onwards to the call to `show`. I think you should gate your choice to use `…` or some syntactically-valid option in `prettyDefForMethod` on this `showForRBI` as well.

You would then at the various call sites of `pretty*ForMethod`, you would choose the options that make the most sense, and they would automatically thread all the way through all the relevant calls.

It may be worth changing that option to say something like "limitToValidSyntax" but that can be done in another change, if at all.

Reply:
And also, you should use my sorbet.run example here as a the start of a new test for your feature.

Reply:
> IMO, we should add concretizeIfAbstract as a boolean to ShowOptions, and then take ShowOptions in this method.

Thanks for the guidance: I've implemented this in 0849d83. That commit also fixes the hover bug you described above.

> It may be worth changing that option to say something like "limitToValidSyntax"

Yes, I thought this was very worth it, since it was otherwise difficult to understand what `showForRBI` meant (especially now that we use that option outside of RBI generation). I opted for the name "useValidSyntax": ced85cc.